ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java

```diff
@@ -225,4 +163,247 @@ private void initMasterKey(String tenantId) {
+            latch.countDown();
+        }
+    }
+
+    private void handlePutDataObjectFailure(
+        String tenantId,
+        String masterKeyId,
+        ThreadContext.StoredContext context,
+        Throwable throwable,
+        AtomicReference<Exception> exceptionRef,
+        CountDownLatch latch
+    ) {
+        Exception cause = SdkClientUtils.unwrapAndConvertToException(throwable, OpenSearchStatusException.class);
+        if (cause instanceof VersionConflictEngineException) {
+            handleVersionConflict(tenantId, masterKeyId, context, exceptionRef, latch);
+        } else {
+            log.debug("Failed to index ML encryption master key to config index", cause);
+            exceptionRef.set(cause);
+            latch.countDown();
+        }
+    }
+
+    private void handleVersionConflict(
+        String tenantId,
+        String masterKeyId,
+        ThreadContext.StoredContext context,
+        AtomicReference<Exception> exceptionRef,
+        CountDownLatch latch
+    ) {
+        sdkClient
+            .getDataObjectAsync(
+                createGetDataObjectRequest(tenantId, new FetchSourceContext(true, Strings.EMPTY_ARRAY, Strings.EMPTY_ARRAY))
+            )
+            .whenComplete(
+                (response, throwable) -> handleVersionConflictResponse(
+                    tenantId,
+                    masterKeyId,
+                    context,
+                    response,
+                    throwable,
+                    exceptionRef,
+                    latch
+                )
+            );
+    }
+
+    private GetDataObjectRequest createGetDataObjectRequest(String tenantId, FetchSourceContext fetchSourceContext) {
+        String masterKeyId = MASTER_KEY;
+        if (tenantId != null) {
+            masterKeyId = MASTER_KEY + "_" + hashString(tenantId);
+        }
+        return GetDataObjectRequest
+            .builder()
+            .index(ML_CONFIG_INDEX)
+            .id(masterKeyId)
+            .tenantId(tenantId)
+            .fetchSourceContext(fetchSourceContext)
+            .build();
+    }
+
+    private void handleVersionConflictResponse(
+        String tenantId,
+        String masterKeyId,
+        ThreadContext.StoredContext context,
+        GetDataObjectResponse response1,
+        Throwable throwable2,
+        AtomicReference<Exception> exceptionRef,
+        CountDownLatch latch
+    ) {
+        context.restore();
+        log.debug("Completed Get config item");
+
+        if (throwable2 != null) {
+            Exception cause1 = SdkClientUtils.unwrapAndConvertToException(throwable2, OpenSearchStatusException.class);
+            log.debug("Failed to get ML encryption master key from config index", cause1);
+            exceptionRef.set(new ResourceNotFoundException(MASTER_KEY_NOT_READY_ERROR));
+            latch.countDown();
+        } else {
+            handleGetDataObjectSuccess(response1, tenantId, masterKeyId, exceptionRef, latch, context);
```

Comment:
I think there is potential dead loop here. In the function `handleGetDataObjectSuccess`, if still getting empty response, the function `initializeNewMasterKey` will be executed again, then `handleVersionConflictResponse` -> `handleGetDataObjectSuccess` -> `initializeNewMasterKey`.
Actually in 2.18, the logic is that if getting `VersionConflictEngineException`, only re-get index once, won't retry creating index.

Reply:
addressed.